### ISSUE TITLE
ext/POSIX/t/sysconf.t:  Unexpected "unopened filehandle" warning

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#18249]
I am getting an unexpected warning when I run file `ext/POSIX/t/sysconf.t` in the core distribution with warnings turned on.

To illustrate the problem, consider the following file in which I have stripped away much of the code from `sysconf.t` to illustrate the current problem.  This file does not need to be run from within a git checkout to demonstrate the problem.  I've been able to reproduce the problem with perl-5.16 and later but not with perl-5.14.
```
$ cat second-sysconf.t 
#!perl -w

BEGIN {
    use Config;
    use Test::More;
    plan skip_all => "POSIX is unavailable" if $Config{'extensions'} !~ m!\bPOSIX\b!;
}

use strict;
use File::Spec;
use POSIX;

sub check(@) {
    grep { eval "&$_;1" or $@!~/vendor has not defined POSIX macro/ } @_
}       

my @path_consts_terminal = check qw(
    _PC_MAX_CANON
);

plan( tests =>  4 );

sub _check_and_report {
    my ($sub, $constant, $description) = @_;
    $! = 0;
    my $return_val = eval {$sub->(eval "$constant()")};
    my $errno = $!; # Grab this before anything else changes it.
    is($@, '', $description);

    # We can't test sysconf further without investigating the type of argument
    # provided
    return if $description =~ /sysconf/;

    if (defined $return_val) {
        like($return_val, qr/\A(?:-?[1-9][0-9]*|0 but true)\z/,
             'the returned value should be a signed integer');
    } else {
      SKIP:
        {
            # POSIX specifies EINVAL is returned if the f?pathconf()
            # isn't implemented for the specific path
            skip "$description not implemented for this path", 1
                if $errno == EINVAL && $description =~ /pathconf/;
            cmp_ok($errno, '==', 0, 'errno should be 0 as before the call')
                or diag("\$!: $errno");
        }
    }
}

my $TTY = "/dev/tty";
SKIP: {
    my $n = 2 * 2 * @path_consts_terminal;

    -c $TTY or skip("$TTY not a character file", $n);
    open(TTY, '<', $TTY) or skip("failed to open $TTY: $!", $n);
    -t TTY or skip("TTY ($TTY) not a terminal file", $n);

    my $fd = fileno(TTY);
    like($fd, qr/^\d+$/, "fileno() returned integer");

    # testing fpathconf() on a terminal file
    for my $constant (@path_consts_terminal) {
        _check_and_report(sub { fpathconf($fd, shift) }, $constant,
            "calling fpathconf($fd, $constant) ($TTY)");
    }
    like($fd, qr/^\d+$/, "fileno() is still an integer");
    
    close($fd);
}
```
When I run this file with a recent perl (note the `-w` on the shebang line), I get:
```
$ perl second-sysconf.t
1..4
ok 1 - fileno() returned integer
ok 2 - calling fpathconf(3, _PC_MAX_CANON) (/dev/tty)
ok 3 - the returned value should be a signed integer
ok 4 - fileno() is still an integer
close() on unopened filehandle 3 at second-sysconf.t line 68.
```
The warning suggests that filehandle `TTY` as been closed somewhere between:
```
    open(TTY, '<', $TTY) or skip("failed to open $TTY: $!", $n);
```
and
```
    close($fd);
```
But I couldn't see where this was happening.

To diagnose, I modified this file to include two unit tests exercising `Scalar::Util::openhandle()`:
```
$ cat third-sysconf.t 
#!perl -w

BEGIN {
    use Config;
    use Test::More;
    plan skip_all => "POSIX is unavailable" if $Config{'extensions'} !~ m!\bPOSIX\b!;
}

use strict;
use File::Spec;
use POSIX;
use Scalar::Util qw( openhandle );

sub check(@) {
    grep { eval "&$_;1" or $@!~/vendor has not defined POSIX macro/ } @_
}       

my @path_consts_terminal = check qw(
    _PC_MAX_CANON
);

plan( tests =>  6 );

sub _check_and_report {
    my ($sub, $constant, $description) = @_;
    $! = 0;
    my $return_val = eval {$sub->(eval "$constant()")};
    my $errno = $!; # Grab this before anything else changes it.
    is($@, '', $description);

    # We can't test sysconf further without investigating the type of argument
    # provided
    return if $description =~ /sysconf/;

    if (defined $return_val) {
        like($return_val, qr/\A(?:-?[1-9][0-9]*|0 but true)\z/,
             'the returned value should be a signed integer');
    } else {
      SKIP:
        {
            # POSIX specifies EINVAL is returned if the f?pathconf()
            # isn't implemented for the specific path
            skip "$description not implemented for this path", 1
                if $errno == EINVAL && $description =~ /pathconf/;
            cmp_ok($errno, '==', 0, 'errno should be 0 as before the call')
                or diag("\$!: $errno");
        }
    }
}

my $TTY = "/dev/tty";
SKIP: {
    my $n = 2 * 2 * @path_consts_terminal;

    -c $TTY or skip("$TTY not a character file", $n);
    open(TTY, '<', $TTY) or skip("failed to open $TTY: $!", $n);
    -t TTY or skip("TTY ($TTY) not a terminal file", $n);

    my $fd = fileno(TTY);
    like($fd, qr/^\d+$/, "fileno() returned integer");
    ok(openhandle($fd), "... and \$fd is open (per Scalar::Util::openhandle)");

    # testing fpathconf() on a terminal file
    for my $constant (@path_consts_terminal) {
        _check_and_report(sub { fpathconf($fd, shift) }, $constant,
            "calling fpathconf($fd, $constant) ($TTY)");
    }
    like($fd, qr/^\d+$/, "fileno() is still an integer");
    ok(openhandle($fd), "... and \$fd is still open (per Scalar::Util::openhandle)");
    
    close($fd);
}
```
The diff between the two files:
```
$ diff -w second-sysconf.t third-sysconf.t 
11a12
> use Scalar::Util qw( openhandle );
21c22
< plan( tests =>  4 );
---
> plan( tests =>  6 );
59a61
>     ok(openhandle($fd), "... and \$fd is open (per Scalar::Util::openhandle)");
66a69
>     ok(openhandle($fd), "... and \$fd is still open (per Scalar::Util::openhandle)");
```
When I run `third-sysconf.t`, I get this:
```
$ perl third-sysconf.t 
1..6
ok 1 - fileno() returned integer
not ok 2 - ... and $fd is open (per Scalar::Util::openhandle)
#   Failed test '... and $fd is open (per Scalar::Util::openhandle)'
#   at third-sysconf.t line 61.
ok 3 - calling fpathconf(3, _PC_MAX_CANON) (/dev/tty)
ok 4 - the returned value should be a signed integer
ok 5 - fileno() is still an integer
not ok 6 - ... and $fd is still open (per Scalar::Util::openhandle)
#   Failed test '... and $fd is still open (per Scalar::Util::openhandle)'
#   at third-sysconf.t line 69.
close() on unopened filehandle 3 at third-sysconf.t line 71.
# Looks like you failed 2 tests of 6.
```
I get failures on each of the two `openhandle()` calls.

Note that the calls to the testing function `_check_and_report`, a wrapper around `POSIX::fpathconf`, do not have an impact on the results.

The documentation for these two functions contains this:
```
$ perldoc -f fileno
    fileno FILEHANDLE
    fileno DIRHANDLE
            Returns the file descriptor for a filehandle or directory
            handle, or undefined if the filehandle is not open. If there is
            no real file descriptor at the OS level, as can happen with
            filehandles connected to memory objects via "open" with a ...

$ perldoc Scalar::Util
...
  openhandle
        my $fh = openhandle( $fh );

    Returns $fh itself, if $fh may be used as a filehandle and is open, or
    if it is a tied handle. Otherwise "undef" is returned.
```
But we know that `fileno` is returning an integer; it's defined and therefore the file cannot be "not open."  We would therefore expect that `openhandle()` would return a true value and the two unit tests would PASS.  They do not PASS.  The fact that an "unopened filehandle" warning is thrown when we attempt to `close($fd);` is consistent with what `openhandle()` is reporting -- but not with `fileno()`.

What is going on here?

Thank you very much.
Jim Keenan

**Comment by Leont** [comment]
`close($fd);` is not valid, `$fd` is a filedescriptor, not a filehandle. It will close the filehandle `*3` instead of the handle `*TTY`. Changing it to `close(TTY)` is the simplest solution, but changing it to use lexical filehandles is probably the better one.

**Comment by jkeenan** [comment]
> `close($fd);` is not valid, `$fd` is a filedescriptor, not a filehandle. It will close the filehandle `*3` instead of the handle `*TTY`. Changing it to `close(TTY)` is the simplest solution, but changing it to use lexical filehandles is probably the better one.

Thanks.  How about https://github.com/Perl/perl5/pull/18250 ?

**Comment by Leont** [comment]
Fixed by #18250
